### ISSUE TITLE
dartanalyzer doesn't work as advertised with @protected

### ISSUE THREAD

**Issue by kendalharland** [dart-lang/sdk#29804]
### protected
The documentation of protected says that it...
>  ...indicates that m should only be invoked on this, whether explicitly or implicitly.

But dartanalyzer does not respect this. For example, no hint or warning is given here:

```dart
library foo;

import 'package:meta/meta.dart';

class Foo {
  @protected
  void protectedMethod() {}

  void violatingMethod() {
    var foo = new Foo();
    foo.protectedMethod(); // Not invoked on an instance of `this`
  }
}
```

### visibleForOverriding
From the docs:
> Used to annotate an instance member that was made public so that it could be overridden but that is not intended to be referenced from outside the defining library.

But this is also not respected by dartanalyzer.  You can basically call a method marked with `visibleForOverriding` wherever you want and you'll never get a hint or warning.  Is this annotation a hangover from some previous change to the analyzer (should it be deleted?) or is this a bug?

**Comment by bwilkerson** [comment]
I absolutely agree that analyzer support would be good to have, but the primary value of the annotations in the 'meta' package is to allow developers to express their intent. There are ways to disable hints such as these, so there isn't any way to prevent users from ignoring the annotations, even if analyzer produced them. Still, analyzer hints would help notify clients that are willing to use code the way it was intended but just didn't notice the annotation, so we should add them when we can.

**Comment by zanderso** [comment]
It would be helpful to have this enforced by the analyzer. I was also led a bit astray by the doc comment in the meta pacakge for this annotation:

> /// Tools, such as the analyzer, can provide feedback if ...

led me to think that not only *could* the analyzer do this, but also that it *would* do this.

**Comment by pq** [comment]
tagging as a `hint` 

@bwilkerson: feel free to update if you'd think this should rather be a **`lint`**?

**Comment by bwilkerson** [comment]
All of the diagnostics related to annotations are implemented as hints, so these should be too.

**Comment by srawlins** [comment]
CC @goderbauer @yjbanov 

The restriction that `@protected` members can only be accessed with a receiver of `this` seems surprising to everyone I've talked to. I suggest that instead of changing analyzer's Hints to match the documentation of  `@protected`, we change the documentation of `@protected` to match analyzer's behavior.

**Comment by yjbanov** [comment]
**tl;dr** I agree with @srawlins that the docs should be updated.

Flutter uses `@protected` to indicate that a method should only be called by the framework. For example, [Widget.createElement](https://api.flutter.dev/flutter/widgets/Widget/createElement.html) and [Widget.createState](https://api.flutter.dev/flutter/widgets/StatefulWidget/createState.html) are implemented by the user, but user code should not call these methods. The framework will call them automatically at the appropriate time in its lifecycle. From where exactly in the framework these methods are called is an implementation detail. Most often it is not the class that defines the protected method ([example](https://github.com/flutter/flutter/blob/a1fa1a3d261f5cee1e8dd2e7a67132f465054d83/packages/flutter/lib/src/widgets/framework.dart#L3292), [example](https://github.com/flutter/flutter/blob/a1fa1a3d261f5cee1e8dd2e7a67132f465054d83/packages/flutter/lib/src/widgets/framework.dart#L4412)), so the "should only be invoked on this" restriction is not useful at all.

Also, it would be weird if `@protected` wasn't a strict superset of private.

**Comment by Hixie** [comment]
"protected" as implemented seems like the right behaviour (visible to library and subclasses).

I would not object to other visibility options like `@strictProtected` (visible to class and subclasses only), or something like `@thisOnly` (visible to this instance only), but I don't think it's a high priority.

**Comment by yjbanov** [comment]
Actually, as implemented, it seems to be incorrect still. It seems `@protected` is not inherited. So if you call `createElement` on the `Widget` type you get analyzer warning, but if you call `createElement` on `StatefulWidget` or `StatelessWidget` you get no warnings.

I think we want `@protected` to be inherited, just like `@immutable` is.

**Comment by DevNico** [comment]
I just stumbled on this issue on my search why @protected doesn't work as documented. Any updates on this issue?

**Comment by srawlins** [comment]
@DevNico apologies; this has been low priority. I think the route we are going to take is to change the documentation of `@protected`, aligning it closer to what the analyzer does.

What specifically isn't working the way you expected?

**Comment by DevNico** [comment]
@srawlins In my use case I have an abstract baseclass containing multiple methods marked with the @protected annotation. Subclasses implement some of the protected methods. I would want the analyzer to throw or rather show an error when methods marked as protected are accessed on instances of the mentioned subclasses.

Thank you for your quick reply.

**Comment by srawlins** [comment]
@DevNico can you give an example where an error is not being reported as you expect. Various things depend on conditions like whether the calling code is in the same library as the `@protected` member, or within a subclass, etc.

It sounds like you want members on subclasses to inherit the `@protected` annotation, like this:

```dart
class C {
  @protected int get a => 7;
}

class D extends C {
  // This should also be considered @protected; accessing `D().a` outside this library should be illegal.
  int get a => 9;
}
```

**Comment by DevNico** [comment]
Almost like you described but basically the same result:

```dart
abstract class C {
  @protected int get a;
}

class D extends C {
  // This should also be considered @protected; accessing `D().a` outside this library should be illegal.
  int get a => 9;
}

void test() {
  final d = D();
  final b = d.a; // This should be illegal
}
```

**Comment by bwilkerson** [comment]
> I think we want `@protected` to be inherited ...

One argument against doing so is that if the annotation is inherited then there isn't any way to make it more visible in a subclass. Not only does that limit the authors of subclasses, but makes it a breaking change to add a protected method to a class that users are allowed to subclass because there might already exist a public method by that name in a subclass that would now become protected.

**Comment by DevNico** [comment]
Inheritance of the protected annotation has it's uses. Maybe another variant of a protected annotation that gets inherited is an option since adding something like that wouldn't be a breaking change but also gives developers more flexibility?

**Comment by DevNico** [comment]
I'd be happy to try and implement such an annotation. I have almost no idea how (yet) but I love me a challenge. If you guys think this is the way to go I would go  ahead and try to implement it. If I fail I'd contact you guys again. Since it's low priority anyway I am guessing this would be alright with you?

**Comment by bwilkerson** [comment]
We need to be thoughtful about the set of annotations we support because every annotation represents a maintenance burden and we, like everyone else, have limited resources. So before we said yes to a new annotation I think we'd want to be certain that we couldn't use the annotation we already have and that there would be enough benefit for users that add it was worth the extra cost to our team.

For example, while I pointed out one argument against making `protected` be inherited, if everyone is already thinking of it as being inherited, and if everyone is likely to switch to a new annotation that _is_ inherited because those are the semantics they want, then there's really no value in having both and we might as well change the `protected` we already have to be what people expect and want it to be. I just don't want us to do that without being aware of the cost of that choice.

**Comment by DevNico** [comment]
That totally makes sense. I feel like the current documentation implies that it would be inherited. I found the `protected` annotation because I was looking for an access modifier similar to Javas protected visibility modifier. Most people probably already think that it is inherited even though that is just my opinion and I don't know how to find out if this is actually true.

IMO it should be inherited. But at the end of the day it's your decision. I am looking forward to you deciding on what to do here!

**Comment by bwilkerson** [comment]
> I found the `protected` annotation because I was looking for an access modifier similar to Java's protected visibility modifier.

Yep, that's what the Dart annotation was modeled after.

But note that (unless things have changed since I last looked at Java)
- In Java the visibility modifiers are not inherited; you always have to specify the visibility you want, even if it's the (no modifier) default of package visible.
- In Java a subclass can override a `protected` method and make it `public`. In Dart we don't have a `public` annotation (because `public` is the default), so there's no way to widen the visibility in a subclass if `protected` is inherited.

Making the `protected` annotation be inherited would make it less like its Java counterpart, which might be confusing to some users.

**Comment by srawlins** [comment]
Note @creativecreatorormaybenot adjusted the text with https://github.com/dart-lang/sdk/commit/9108f927ab92c6d679f551d1fb4a4d889c3b8546.

**Comment by srawlins** [comment]
I'm forking off the `@visibleForOverriding` issue into its own. https://github.com/dart-lang/sdk/issues/46155
